util/project_allocation_mapper.py

```diff
@@ -0,0 +1,509 @@
+        charge_code = None
+        if publication.project_id:
+            try:
+                project = portal_proj.objects.get(pk=publication.project_id)
+                nickname = project.nickname
+                charge_code = project.charge_code
+            except ProjectExtras.DoesNotExist:
+                logger.warning('Couldn\'t find project with id {} in portal'.format(publication.project_id))
+        elif publication.tas_project_id:
+            pextras = ProjectExtras.objects.filter(tas_project_id=publication.tas_project_id)
+            if pextras and pextras.count() > 0:
+                nickname = pextras[0].nickname
+            charge_code = publication.tas_project_id
+        
+        return nickname, charge_code
+        
+    def get_user_projects(self, username, to_pytas_model=False):
+        user_projects = []
+        if self.is_from_db:
+            keycloak_client = KeycloakClient()
+            for proj in keycloak_client.get_user_projects_by_username(username):
+                project = portal_proj.objects.filter(charge_code=proj)
+                if len(project) == 0:
+                    logger.error('Couldn\'t find project {} in portal'.format(proj))
+                else:
+                    project = self.portal_to_tas_proj_obj(project[0], to_pytas_model=to_pytas_model)
+                    project_allocs = self._get_project_allocations(project, to_pytas_model=to_pytas_model)
+                    self.set_attr(project, 'allocations', project_allocs)
+                    user_projects.append(project)
+        else:
+            if to_pytas_model:
+                user_projects = tas_proj.list(username=username)
+            else:
+                user_projects = self.tas.projects_for_user(username=username)
+                        
+            for proj in user_projects:
+                try:
+                    extras = ProjectExtras.objects.get(tas_project_id=self.get_attr(proj, 'id'))
+                    try:
+                        proj.__dict__['nickname'] = extras.nickname
+                    except:
+                        proj['nickname'] = extras.nickname
+                except ProjectExtras.DoesNotExist:
+                    project_nickname = None
+            
+        return user_projects
+    
+    def get_project_members(self, tas_project):
+        users = []
+        if self.is_from_db:
+            keycloak_client = KeycloakClient()
+            pi_username = tas_project.pi.username
+            for username in keycloak_client.get_project_members_by_charge_code(tas_project.chargeCode):
+                # assign role
+                role = 'Standard'
+                if username == pi_username:
+                    role = 'PI'
+                users.append(self.portal_keycloak_user_to_tas_obj(username, to_pytas_model=True, role=role))
+        else:
+            users = tas_project.get_users()
+        return users
+    
+    def get_project(self, project_id):
```

Comment:
This function would be where you'd first try to fetch the project from the DB and then fall back to TAS, before giving up entirely.